pkg/tasks/task_action_builder.go

```diff
@@ -0,0 +1,126 @@
+package tasks
+
+import (
+	"path/filepath"
+
+	"github.com/devbuddy/devbuddy/pkg/helpers/store"
+	"github.com/devbuddy/devbuddy/pkg/utils"
+)
+
+type actionRunFunc func(*context) error
+
+type actionCondition struct {
+	pre  func(*context) *actionResult
+	post func(*context) *actionResult
+}
+
+type actionWithBuilder struct {
+	desc       string
+	conditions []*actionCondition
+	runFunc    actionRunFunc
+	ran        bool
+}
+
+func (a *actionWithBuilder) description() string {
+	return a.desc
+}
+
+func (a *actionWithBuilder) needed(ctx *context) (result *actionResult) {
+	if a.ran {
+		return a.post(ctx)
+	}
+	return a.pre(ctx)
+}
+
+func (a *actionWithBuilder) pre(ctx *context) (result *actionResult) {
+	if len(a.conditions) == 0 {
+		return actionNeeded("")
+	}
+	for _, condition := range a.conditions {
+		result = condition.pre(ctx)
+		if result.Error != nil || result.Needed {
+			return result
+		}
+	}
+	return actionNotNeeded()
+}
+
+func (a *actionWithBuilder) post(ctx *context) (result *actionResult) {
+	for _, condition := range a.conditions {
+		result = condition.post(ctx)
+		if result.Error != nil || result.Needed {
+			return result
+		}
+	}
+	return actionNotNeeded()
+}
+
+func (a *actionWithBuilder) run(ctx *context) error {
+	a.ran = true
+	return a.runFunc(ctx)
+}
+
+// func (a *actionWithBuilder) setDescription(desc string) *actionWithBuilder {
+// 	a.desc = desc
+// 	return a
+// }
+
+func (a *actionWithBuilder) addCondition(condition *actionCondition) *actionWithBuilder {
+	a.conditions = append(a.conditions, condition)
+	return a
+}
+
+func (a *actionWithBuilder) addConditionFunc(condFunc func(*context) *actionResult) *actionWithBuilder {
```

Comment:
I'm wondering if it make sense to pretend those functions are "private" since they are part of the Task api `addCondition` is clearly internal but `addConditionFunc` is part of the "public" one in our case in a way?